packages/api/lib/ems.js

```diff
@@ -227,71 +39,47 @@ function buildReportFileName(reportType, startTime) {
+async function determineReportKey(reportType, reportStartTime, reportsPrefix) {
+  let reportName = buildReportFileName(reportType, reportStartTime);
 
-  if (!Object.keys(emsMappings).includes(reportType)) {
-    throw new Error(`ems.generateReport report type not supported: ${reportType}`);
-  }
+  const revisionNumber = (await aws.listS3ObjectsV2({
+    Bucket: process.env.system_bucket,
+    Prefix: aws.s3Join([reportsPrefix, reportName])
+  })).length;
 
-  // create a temporary file for the report
-  const name = buildReportFileName(reportType, startTime);
-  const filename = path.join(os.tmpdir(), name);
-  const stream = fs.createWriteStream(filename);
+  if (revisionNumber > 0) reportName = `${reportName}.rev${revisionNumber}`;
 
-  // retrieve granule/deletedgranule records in batches, and generate EMS records for each batch
-  const esClient = await Search.es();
-  const type = (reportType !== 'delete') ? 'granule' : 'deletedgranule';
-
-  const esIndex = process.env.ES_INDEX || defaultIndexAlias;
-  const searchQuery = buildSearchQuery(esIndex, type, startTime, endTime);
-  let response = await esClient.search(searchQuery);
-  let granules = response.hits.hits.map((s) => s._source);
-  let numRetrieved = granules.length;
-  stream.write(buildEMSRecords(emsMappings[reportType], granules).join('\n'));
-
-  while (response.hits.total !== numRetrieved) {
-    response = await esClient.scroll({ // eslint-disable-line no-await-in-loop
-      scrollId: response._scroll_id,
-      scroll: '30s'
-    });
-    granules = response.hits.hits.map((s) => s._source);
-    stream.write('\n');
-    stream.write(buildEMSRecords(emsMappings[reportType], granules).join('\n'));
-    numRetrieved += granules.length;
-  }
-  stream.end();
-  log.debug(`EMS ${reportType} generated with ${numRetrieved} records: ${filename}`);
-
-  // upload to s3
-  const s3Uri = await uploadReportToS3(filename);
-  return { reportType, file: s3Uri };
+  return aws.s3Join([reportsPrefix, reportName]);
 }
 
 /**
- * generate all EMS reports given the time range of the records and submit to ems
- *
- * @param {string} startTime - start time of the records
- * @param {string} endTime - end time of the records
- * @returns {Array<Object>} - list of report type and its file path {reportType, file}
+ * get list of expired s3 objects
+ * @param {string} bucket - the s3 bucket
+ * @param {string} prefix - the S3 prefix where the objects are located
+ * @param {string} retentionInDays - the retention in days for the s3 objects
+ * @returns {Array<Object>} - list of s3 objects
  */
-async function generateReports(startTime, endTime) {
-  return Promise.all(Object.keys(emsMappings)
-    .map((reportType) => generateReport(reportType, startTime, endTime)));
+async function getExpiredS3Objects(bucket, prefix, retentionInDays) {
+  const retentionFilter = (s3Object) =>
+    s3Object.LastModified <= moment.utc().subtract(retentionInDays, 'days');
```

Comment:
The aws sdk documentation is not clear on this - is the `LastModified` value returned by `listS3ObjectsV2` comparable with a `moment` time?

Reply:
Testing locally (sorry for the terrible formatting)...
```
> aws.listS3ObjectsV2(params).then((response) => response.map((r) => console.log(`${r.LastModified} <= ${moment.utc().subtract(100, 'days')}?`, r.LastModified <= moment.utc().subtract(100, 'days'))));
```
returns:
```
Mon Mar 25 2019 10:12:35 GMT-0500 (CDT) <= Tue Mar 05 2019 22:35:25 GMT+0000? false
Mon Mar 25 2019 10:12:35 GMT-0500 (CDT) <= Tue Mar 05 2019 22:35:25 GMT+0000? false
Mon Mar 25 2019 10:12:35 GMT-0500 (CDT) <= Tue Mar 05 2019 22:35:25 GMT+0000? false
Mon Mar 25 2019 10:12:35 GMT-0500 (CDT) <= Tue Mar 05 2019 22:35:25 GMT+0000? false
Mon Mar 25 2019 10:12:35 GMT-0500 (CDT) <= Tue Mar 05 2019 22:35:25 GMT+0000? false
Mon Mar 25 2019 10:12:35 GMT-0500 (CDT) <= Tue Mar 05 2019 22:35:25 GMT+0000? false
Thu Dec 13 2018 16:09:19 GMT-0600 (CST) <= Tue Mar 05 2019 22:35:25 GMT+0000? true
Fri Feb 08 2019 11:48:32 GMT-0600 (CST) <= Tue Mar 05 2019 22:35:25 GMT+0000? true
Thu Dec 13 2018 12:59:54 GMT-0600 (CST) <= Tue Mar 05 2019 22:35:25 GMT+0000? true
Thu Dec 13 2018 12:44:31 GMT-0600 (CST) <= Tue Mar 05 2019 22:35:25 GMT+0000? true
Thu Dec 13 2018 16:38:49 GMT-0600 (CST) <= Tue Mar 05 2019 22:35:25 GMT+0000? true
Thu Dec 13 2018 15:57:31 GMT-0600 (CST) <= Tue Mar 05 2019 22:35:25 GMT+0000? true
Thu Dec 13 2018 15:50:31 GMT-0600 (CST) <= Tue Mar 05 2019 22:35:25 GMT+0000? true
Thu Dec 13 2018 16:15:13 GMT-0600 (CST) <= Tue Mar 05 2019 22:35:25 GMT+0000? true
```

Which looks _almost_ good, but there is a mis-match there in the timezone that worries me.

Reply:
Good change :+1:. According to https://stackoverflow.com/questions/7606798/javascript-date-object-comparison it looks like an `==` comparison will never resolve to true because the code is comparing two objects.

I think it would make sense to either change `<=` to `<` or, as suggested in the above link, change `toDate()` to `toTime()`:

`s3Object.LastModified.getTime() <= moment.utc().subtract(retentionInDays, 'days').toDate().getTime();`